components/locid/src/extensions/mod.rs

```diff
@@ -135,6 +135,45 @@ impl Extensions {
             && self.other.is_empty()
     }
 
+    /// Retains the specified extension types, clearing all others.
+    ///
+    /// # Examples
+    ///
+    /// ```
+    /// use std::str::FromStr;
+    /// use icu::locid::Locale;
+    /// use icu::locid::extensions::ExtensionType;
+    ///
+    /// let loc: Locale = "und-a-hello-t-mul-u-world-z-zzz-x-extra".parse().unwrap();
+    ///
+    /// let mut only_unicode = loc.clone();
+    /// only_unicode.extensions.retain_by_type(|t| t == ExtensionType::Unicode);
+    /// assert_eq!(only_unicode, "und-u-world");
+    ///
+    /// let mut only_t_z = loc.clone();
+    /// only_t_z.extensions.retain_by_type(|t| {
+    ///     t == ExtensionType::Transform
+    ///         || t == ExtensionType::Other(b'z')
+    /// });
+    /// assert_eq!(only_t_z, "und-t-mul-z-zzz");
+    /// ```
+    pub fn retain_by_type<F>(&mut self, mut predicate: F)
+    where
+        F: FnMut(ExtensionType) -> bool,
```

Comment:
`suggestion`: it seems that you don't need this to be mutable - you can use `Fn` here. What's more, you can just use `predicate: fn(usize) -> bool`.

Reply:
I based this signature on `Vec::retain`

https://doc.rust-lang.org/std/vec/struct.Vec.html#method.retain

> `fn(usize) -> bool`

what did you envision would be in the `usize`?

Reply:
CC @Manishearth

Reply:
In general for closures it's helpful to use FnMut where possible for maximum flexibility. Using FnMut is not constraining the user of this API, it constrains us (as implementors), wheras using Fn/FnOnce constrain the user while giving us more freedom.

Reply:
gotcha. thanks!